cdap-app-fabric/src/main/java/io/cdap/cdap/gateway/handlers/AppLifecycleHttpHandler.java

```diff
@@ -506,6 +510,8 @@ public void upgradeApplication(HttpRequest request, HttpResponder responder,
                                  @QueryParam("artifactScope") Set<String> artifactScopes,
                                  @QueryParam("allowSnapshot") boolean allowSnapshot) throws Exception {
     ApplicationId appId = validateApplicationId(validateNamespace(namespaceId), appName);
+    String latestAppVersion = getLatestAppVersion(appId);
```

Comment:
Same race condition as above. Please open a JIRA and we need to see how to handle it correctly.

Reply:
yes agree - I was skeptical about this one too, updated this JIRA and will remove this change https://cdap.atlassian.net/browse/CDAP-19782

Reply:
pulled this out